### PR TITLE
Downgrade cri-o to 1.11.6

### LINKED ISSUE #746: Kontena-storage broken on cri-o
```
2018-11-01 05:39:24.123471 I | exec: Running command: lsblk --all --noheadings --list --output KNAME
2018-11-01 05:39:24.123555 I | rook-discover: failed to probe devices: failed initial hardware discovery. failed to list all devices: Failed to complete 'lsblk all': open /dev/null: permission denied. 
2018-11-01 05:39:24.123568 I | rook-discover: failed to update device configmap: failed initial hardware discovery. failed to list all devices: Failed to complete 'lsblk all': open /dev/null: permission denied. 
failed initial hardware discovery. failed to list all devices: Failed to complete 'lsblk all': open /dev/null: permission denied. 
```

Downgraded cri-o to 1.11.6 and it seems to work fine.

### PR DESCRIPTION
**Pull request by jakolehm**
Downgrades cri-o to v1.11.6.

Downgrade revealed interesting side-effects.. that's why this PR also improves container runtime upgrade/downgrade by making it more safe:
- drain/uncordon node if master is healthy
- allow only new nodes to be configured parallel

Fixes #746

### CODE REVIEW

File: lib/pharos/phases/configure_host.rb [hunk 1]
```diff
@@ -23,8 +25,31 @@ def call
...
+            end
+          }
+        end
+      end
+
+      def master_ssh
+        ssh_manager.client_for(@master)
+      end
+
+      def master_healthy?
```
Comment (jakolehm) [inline lib/pharos/phases/configure_host.rb]: Not sure if this method actually should go to the `Pharos::Configuration::Host`?

File: lib/pharos/ssh/manager.rb [hunk 2]
```diff
@@ -6,6 +6,8 @@
 module Pharos
   module SSH
     class Manager
+      include Singleton
```
Comment (jakolehm) [inline lib/pharos/ssh/manager.rb]: We need just one manager.

File: lib/pharos/host/ubuntu/scripts/configure-cri-o.sh [hunk 3]
```diff
@@ -80,11 +78,18 @@ if ! systemctl is-active --quiet crio; then
     systemctl enable crio
     systemctl start crio
 else
+    if [ -f /etc/cni/net.d/100-crio-bridge.conf ] || [ -f /etc/cni/net.d/200-loopback.conf ]; then
```
Comment (jakolehm) [inline lib/pharos/host/ubuntu/scripts/configure-cri-o.sh]: Need to add this to el7 also.

**Comment by jakolehm** [comment]
Example reconfigure output:
```
==> Configure hosts @ jari-master-0 jari-worker-0 jari-worker-1 jari-worker-2
    [jari-master-0] Configuring script helpers ...
    [jari-worker-2] Configuring script helpers ...
    [jari-worker-1] Configuring script helpers ...
    [jari-worker-0] Configuring script helpers ...
    [jari-worker-0] Configuring essential packages ...
    [jari-master-0] Configuring essential packages ...
    [jari-worker-1] Configuring essential packages ...
    [jari-worker-2] Configuring essential packages ...
    [jari-worker-0] Configuring package repositories ...
    [jari-worker-2] Configuring package repositories ...
    [jari-master-0] Configuring package repositories ...
    [jari-worker-1] Configuring package repositories ...
    [jari-worker-0] Configuring netfilter ...
    [jari-worker-0] Draining node ...
    [jari-worker-2] Configuring netfilter ...
    [jari-master-0] Configuring netfilter ...
    [jari-worker-1] Configuring netfilter ...
    [jari-worker-0] Reconfiguring container runtime (cri-o) packages ...
    [jari-worker-0] Uncordoning node ...
    [jari-worker-2] Draining node ...
    [jari-worker-2] Reconfiguring container runtime (cri-o) packages ...
    [jari-worker-2] Uncordoning node ...
    [jari-master-0] Draining node ...
    [jari-master-0] Reconfiguring container runtime (cri-o) packages ...
    [jari-master-0] Uncordoning node ...
    [jari-worker-1] Draining node ...
    [jari-worker-1] Reconfiguring container runtime (cri-o) packages ...
    [jari-worker-1] Uncordoning node ...
```

File: lib/pharos/phases/configure_host.rb [hunk 4]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'singleton'
```
Comment (kke) [inline lib/pharos/phases/configure_host.rb]: Why is this here?

File: lib/pharos/phases/configure_host.rb [hunk 5]
```diff
@@ -23,8 +25,31 @@ def call
         logger.info { "Configuring netfilter ..." }
         host_configurer.configure_netfilter
 
-        logger.info { "Configuring container runtime (#{@host.container_runtime}) packages ..." }
-        host_configurer.configure_container_runtime
+        if @host.new?
+          logger.info { "Configuring container runtime (#{@host.container_runtime}) packages ..." }
+          host_configurer.configure_container_runtime
+        else
+          mutex.synchronize {
```
Comment (kke) [inline lib/pharos/phases/configure_host.rb]: Where does mutex come from?

File: lib/pharos/phases/configure_host.rb (hunk 4 above)
Comment (jakolehm) [reply lib/pharos/phases/configure_host.rb]: 🙈 , fixed.

File: lib/pharos/phases/configure_host.rb (hunk 5 above)
Comment (jakolehm) [reply lib/pharos/phases/configure_host.rb]: From the Phase class (was missing).

File: lib/pharos/phases/configure_host.rb [hunk 6]
```diff
@@ -23,8 +23,31 @@ def call
...
```
Comment (kke) [inline lib/pharos/phases/configure_host.rb]:
```suggestion
          mutex.synchronize do
```

File: lib/pharos/phases/configure_host.rb [hunk 7]
```diff
@@ -23,8 +23,31 @@ def call
...
+              logger.info { "Draining node ..." }
+              master_ssh.exec!("kubectl drain --force --ignore-daemonsets --delete-local-data #{@host.hostname}")
+            end
+            logger.info { "Reconfiguring container runtime (#{@host.container_runtime}) packages ..." }
+            host_configurer.configure_container_runtime
+            if master_healthy?
+              logger.info { "Uncordoning node ..." }
+              master_ssh.exec!("kubectl uncordon #{@host.hostname}")
+            end
+          }
```
Comment (kke) [inline lib/pharos/phases/configure_host.rb]:
```suggestion
          end
```

File: lib/pharos/host/ubuntu/scripts/configure-cri-o.sh [hunk 8]
```diff
@@ -62,11 +62,9 @@ fi
 
 export DEBIAN_FRONTEND=noninteractive
 apt-mark unhold cri-o
-apt-get install -y cri-o="${CRIO_VERSION}"
+apt-get install -y --allow-downgrades cri-o="${CRIO_VERSION}"
```
Comment (jnummelin) [inline lib/pharos/host/ubuntu/scripts/configure-cri-o.sh]: I guess we need this flag also on other distros

File: lib/pharos/phases/configure_host.rb (hunk 1 above)
Comment (kke) [reply lib/pharos/phases/configure_host.rb]: That requires that hosts know who their master is

Comment (kke) [reply lib/pharos/phases/configure_host.rb]: Or you mean it should be a method for master hosts so you can ask if it thinks it's healthy? Could very well be there, yes.

File: lib/pharos/host/ubuntu/scripts/configure-cri-o.sh (hunk 8 above)
Comment (jakolehm) [reply lib/pharos/host/ubuntu/scripts/configure-cri-o.sh]: Fixed.

**Comment by jakolehm** [comment]
Now phase outsources the decision making process to OS configurer... which means that only bionic will do a rolling reconfigure (others can reconfigure without stopping pods).

**Comment by jakolehm** [comment]
@jnummelin PTAL

File: lib/pharos/host/el7/scripts/configure-essentials.sh [hunk 9]
```diff
@@ -12,6 +12,9 @@ yum_install_with_lock() {
     version=$2
     linefromfile "^0:${package}-" $versionlock
     yum install -y "${package}-${version}"
+    if ! rpm -qi "${package}-${version}" > /dev/null ; then
```
Comment (jnummelin) [inline lib/pharos/host/el7/scripts/configure-essentials.sh]: I do not get how this works 🤔 Does `yum install ...` just fail silently if trying to install older package version?

File: lib/pharos/host/ubuntu/ubuntu_bionic.rb [hunk 10]
```diff
@@ -56,6 +56,22 @@ def configure_container_runtime
           raise Pharos::Error, "Unknown container runtime: #{host.container_runtime}"
         end
       end
+
```
Comment (jnummelin) [inline lib/pharos/host/ubuntu/ubuntu_bionic.rb]: Both Xenial and Bionic have the exact same logic, could this method come from their common super class?

File: lib/pharos/host/el7/el7.rb [hunk 11]
```diff
@@ -69,6 +69,10 @@ def configure_container_runtime
         end
       end
 
+      def configure_container_runtime_safe?
+        true
```
Comment (jnummelin) [inline lib/pharos/host/el7/el7.rb]: Why is EL based distros always safe?

Comment (jakolehm) [reply lib/pharos/host/el7/el7.rb]: I didn't yet come up with logic how to resolve versions etc.. it's not worse than the current version.

Comment (jakolehm) [reply lib/pharos/host/el7/el7.rb]: Fixed.

File: lib/pharos/host/el7/scripts/configure-essentials.sh (hunk 9 above)
Comment (jakolehm) [reply lib/pharos/host/el7/scripts/configure-essentials.sh]: Yes.

File: lib/pharos/host/ubuntu/ubuntu_bionic.rb (hunk 10 above)
Comment (jakolehm) [reply lib/pharos/host/ubuntu/ubuntu_bionic.rb]: Fixed.

**Comment by jakolehm** [comment]
@jnummelin PTAL again

**Pull request merged**